services/core/shared/database/knex.js

```diff
@@ -0,0 +1,135 @@
+ * propagated, or distributed except according to the terms contained in the
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ *
+ */
+const { Logger } = require('lisk-service-framework');
+
+const logger = Logger();
+
+const createDb = async () => {
+    const knex = require('knex')({
+        client: 'mysql',
+        version: '5.7',
+        connection: {
+            host: '127.0.0.1',
+            user: 'lisk',
+            password: 'password',
+            database: 'lisk',
+        },
+        // useNullAsDefault: true,
+        log: {
+            warn(message) { logger.warn(message); },
+            error(message) { logger.error(message); },
+            deprecate(message) { logger.deprecate(message); },
+            debug(message) { logger.debug(message); },
+        },
+        migrations: {
+            directory: './database/knex_migrations',
+            loadExtensions: ['.js'],
+        },
+    });
+
+    await knex.migrate.latest();
+
+    return knex;
+};
+
+let dbConnection = null;
+
+const getDbInstance = async (tableName) => {
+    if (!dbConnection) dbConnection = await createDb();
+
+    const knex = dbConnection;
+
+    const write = async (row) => knex.transaction(async trx => {
+        const inserts = await trx(tableName).insert(row).onConflict('id').merge()
+            .transacting(trx);
+        logger.info(`${inserts.length} row inserted/updated in '${tableName}' table`);
+        return inserts;
+    });
+
+    const writeOnce = async (row) => write(row);
+
+    const writeBatch = async (rows) => {
+        try {
+            const chunkSize = 1000;
+            const ids = await knex.batchInsert(tableName, rows, chunkSize);
+            logger.info(`${rows.length} row(s) inserted/updated in '${tableName}' table`);
+            return ids;
+        } catch (error) {
+            const errCode = error.code;
+            const errMessage = error.message.split(`${errCode}: `)[1] || error.message;
+            const errCause = errMessage.split(': ')[0];
+            logger.debug('Encountered error with batch insert:', errCause,
+                '\nRe-attempting to update/merge the conflicted transactions one at a time: ');
+
+            return knex.transaction(async trx => {
+                const inserts = await Promise.all(rows.map(row => trx(tableName).insert(row).onConflict('id').merge()
+                    .transacting(trx)));
+                logger.info(`${rows.length} row(s) inserted/updated in '${tableName}' table`);
+                return inserts;
+            });
+        }
+    };
+
+    const findAll = async () => knex.select().table(tableName);
+
+    const find = async (params) => {
+        const whereParams = params.selector;
+        const res = await knex.select().table(tableName).where(whereParams);
+        return res;
+    };
+
+    const findById = async (id) => {
+        const params = { selector: { id } };
```

Comment:
Remove `selector` as it is specific to PouchDB.

Reply:
Done!